File: SundeeFundee/Domain/CycleCalculations.swift
import Foundation

// MARK: - Shared enums

enum CyclePhase: String, Codable, CaseIterable {
    case menstrual, follicular, ovulation, luteal
}

// MARK: - Result types

struct CycleStatusResult {
    let currentPhase: CyclePhase
    let cycleDay: Int
    let daysUntilNextPhase: Int
    let predictedNextPeriod: Date
    let phaseStartDate: Date
    let phaseEndDate: Date
}

struct PhaseBoundary {
    let start: Int
    let end: Int
}

struct PhaseRecommendation {
    let phase: CyclePhase
    let title: String
    let description: String
    let trainingFocus: String
    let intensityRecommendation: String
    let exercisesToEmphasize: [String]
    let exercisesToAvoid: [String]
}

// MARK: - CycleCalculations

enum CycleCalculations {

    /// Returns nil if no period logs are available.
    static func calculateCycleStatus(
        periodLogs: [PeriodLog],
        settings: CycleSettings,
        referenceDate: Date = .now
    ) -> CycleStatusResult? {
        guard !periodLogs.isEmpty else { return nil }

        let ref = startOfDay(referenceDate)
        let sorted = periodLogs.sorted { $0.startDate > $1.startDate }

        var cycleStartDate: Date?

        for period in sorted {
            let pStart = startOfDay(period.startDate)
            let pEnd   = period.endDate.map { startOfDay($0) }
                      ?? addDays(pStart, settings.averagePeriodLengthDays - 1)

            if isWithin(ref, start: pStart, end: pEnd) {
                cycleStartDate = pStart; break
            }
            let nextExpected = addDays(pStart, settings.averageCycleLengthDays)
            if ref > pEnd && ref < nextExpected {
                cycleStartDate = pStart; break
            }
        }

        let cycleStart: Date
        if let unwrapped = cycleStartDate {
            cycleStart = unwrapped
        } else {
            let most  = sorted[0]
            var start = startOfDay(most.startDate)
            let daysSince = daysBetween(start, ref)
            let completed = daysSince / settings.averageCycleLengthDays
            start = addDays(start, completed * settings.averageCycleLengthDays)
            cycleStart = start
        }

        let cycleDay  = daysBetween(cycleStart, ref) + 1
        let boundaries = getPhaseBoundaries(settings: settings)

        var currentPhase  = CyclePhase.follicular
        var phaseStartDay = 1
        var phaseEndDay   = settings.averageCycleLengthDays

        for phase in CyclePhase.allCases {
            let b = boundaries[phase]!
            if cycleDay >= b.start && cycleDay <= b.end {
                currentPhase  = phase
                phaseStartDay = b.start
                phaseEndDay   = b.end
                break
            }
        }

        let daysUntilNext: Int
        switch currentPhase {
        case .menstrual:
            daysUntilNext = boundaries[.follicular]!.start - cycleDay
        case .follicular:
            daysUntilNext = boundaries[.ovulation]!.start - cycleDay
        case .ovulation:
            daysUntilNext = boundaries[.luteal]!.start - cycleDay
        case .luteal:
            daysUntilNext = settings.averageCycleLengthDays - cycleDay + 1
        }

        return CycleStatusResult(
            currentPhase:       currentPhase,
            cycleDay:           cycleDay,
            daysUntilNextPhase: max(0, daysUntilNext),
            predictedNextPeriod: addDays(cycleStart, settings.averageCycleLengthDays),
            phaseStartDate:      addDays(cycleStart, phaseStartDay - 1),
            phaseEndDate:        addDays(cycleStart, phaseEndDay   - 1)
        )
    }

    static func getPhaseBoundaries(settings: CycleSettings) -> [CyclePhase: PhaseBoundary] {
        let cycleLen   = settings.averageCycleLengthDays
        let periodLen  = settings.averagePeriodLengthDays
        let lutealLen  = settings.lutealPhaseLengthDays

        let ovDay        = cycleLen - lutealLen
        let ovStart      = max(periodLen + 2, ovDay - 2)
        let ovEnd        = min(ovDay + 2, cycleLen - lutealLen + 2)

        return [
            .menstrual:  PhaseBoundary(start: 1,              end: periodLen),
            .follicular: PhaseBoundary(start: periodLen + 1,  end: ovStart - 1),
            .ovulation:  PhaseBoundary(start: ovStart,        end: ovEnd),
            .luteal:     PhaseBoundary(start: ovEnd + 1,      end: cycleLen),
        ]
    }

    static func getPhaseRecommendation(_ phase: CyclePhase) -> PhaseRecommendation {
        switch phase {
        case .menstrual:
            return PhaseRecommendation(
                phase: .menstrual,
                title: "Menstrual Phase",
                description: "Your period phase. Energy may be lower — you might feel more fatigued.",
                trainingFocus: "Recovery and light movement",
                intensityRecommendation: "low",
                exercisesToEmphasize: ["yoga", "walking", "light stretching"],
                exercisesToAvoid: ["heavy compound lifts", "max effort attempts"]
            )
        case .follicular:
            return PhaseRecommendation(
                phase: .follicular,
                title: "Follicular Phase",
                description: "Energy and endurance begin to rise. Estrogen increases, supporting muscle growth.",
                trainingFocus: "Building strength and endurance",
                intensityRecommendation: "moderate",
                exercisesToEmphasize: ["compound movements", "strength training", "cardio"],
                exercisesToAvoid: []
            )
        case .ovulation:
            return PhaseRecommendation(
                phase: .ovulation,
                title: "Ovulation Phase",
                description: "Peak estrogen and testosterone. Often the strongest phase for performance.",
                trainingFocus: "High-intensity training and PR attempts",
                intensityRecommendation: "peak",
                exercisesToEmphasize: ["max effort attempts", "heavy compound lifts", "power-focused workouts"],
                exercisesToAvoid: []
            )
        case .luteal:
            return PhaseRecommendation(
                phase: .luteal,
                title: "Luteal Phase",
                description: "Progesterone rises, which may affect recovery and energy. Focus on maintenance.",
                trainingFocus: "Maintenance and technique refinement",
                intensityRecommendation: "moderate",
                exercisesToEmphasize: ["technique work", "volume training", "recovery-focused sessions"],
                exercisesToAvoid: ["max effort attempts", "extremely heavy loads"]
            )
        }
    }

    // MARK: - Date helpers

    private static func startOfDay(_ date: Date) -> Date {
        Calendar.current.startOfDay(for: date)
    }

    private static func addDays(_ date: Date, _ days: Int) -> Date {
        Calendar.current.date(byAdding: .day, value: days, to: startOfDay(date))!
    }

    private static func daysBetween(_ from: Date, _ to: Date) -> Int {
        Calendar.current.dateComponents([.day], from: startOfDay(from), to: startOfDay(to)).day!
    }

    private static func isWithin(_ target: Date, start: Date, end: Date) -> Bool {
        target >= start && target <= end
    }
}

// MARK: - CyclePhase display helpers

extension CyclePhase {
    var displayName: String {
        switch self {
        case .menstrual:  return "Menstrual"
        case .follicular: return "Follicular"
        case .ovulation:  return "Ovulation"
        case .luteal:     return "Luteal"
        }
    }

    var emoji: String {
        switch self {
        case .menstrual:  return "🌊"
        case .follicular: return "🌱"
        case .ovulation:  return "⚡️"
        case .luteal:     return "🌕"
        }
    }
}

File: src/domain/cycle/cycle-calculations.ts
// =============================================================================
// CycleCalculations — TypeScript port of CycleCalculations.swift
// Uses date-fns for all calendar arithmetic (exact Swift Calendar.current parity).
// Pure functions — no side effects.
// =============================================================================

import { startOfDay, differenceInCalendarDays, addDays } from 'date-fns';
import type { CyclePhase, CycleSettings, PeriodLog } from '../types';

// ---------------------------------------------------------------------------
// Result types
// ---------------------------------------------------------------------------

export interface PhaseBoundary {
  start: number;
  end: number;
}

export interface PhaseBoundaries {
  menstrual: PhaseBoundary;
  follicular: PhaseBoundary;
  ovulation: PhaseBoundary;
  luteal: PhaseBoundary;
}

export interface CycleStatusResult {
  currentPhase: CyclePhase;
  cycleDay: number;
  daysUntilNextPhase: number;
  predictedNextPeriod: Date;
  phaseStartDate: Date;
  phaseEndDate: Date;
}

// ---------------------------------------------------------------------------
// Phase boundaries
// ---------------------------------------------------------------------------

/**
 * Compute start/end cycle-day boundaries for each phase given cycle settings.
 * Mirrors CycleCalculations.getPhaseBoundaries(_:) exactly.
 */
export function getPhaseBoundaries(settings: CycleSettings): PhaseBoundaries {
  const cycleLen = settings.averageCycleLengthDays;
  const periodLen = settings.averagePeriodLengthDays;
  const lutealLen = settings.lutealPhaseLengthDays;

  const ovDay = cycleLen - lutealLen;
  const ovStart = Math.max(periodLen + 2, ovDay - 2);
  const ovEnd = Math.max(ovStart, Math.min(ovDay + 2, cycleLen - lutealLen + 2));
  const follicularEnd = Math.max(periodLen + 1, ovStart - 1);

  return {
    menstrual:  { start: 1,            end: periodLen     },
    follicular: { start: periodLen + 1, end: follicularEnd },
    ovulation:  { start: ovStart,       end: ovEnd         },
    luteal:     { start: ovEnd + 1,     end: cycleLen      },
  };
}

// ---------------------------------------------------------------------------
// Phase inference
// ---------------------------------------------------------------------------

/** Ordered cycle phases matching Swift CyclePhase.allCases */
const PHASE_ORDER: CyclePhase[] = ['menstrual', 'follicular', 'ovulation', 'luteal'];

/**
 * Infer the current cycle phase from a cycle-day number and settings.
 * Mirrors Swift's phase lookup loop in calculateCycleStatus.
 * Falls back to 'follicular' if no boundary matches (parity with Swift default).
 */
export function inferCurrentPhase(cycleDay: number, settings: CycleSettings): CyclePhase {
  const boundaries = getPhaseBoundaries(settings);
  for (const phase of PHASE_ORDER) {
    const b = boundaries[phase];
    if (cycleDay >= b.start && cycleDay <= b.end) {
      return phase;
    }
  }
  return 'follicular';
}

// ---------------------------------------------------------------------------
// Date helpers (private)
// ---------------------------------------------------------------------------

function daysBetween(from: Date, to: Date): number {
  return differenceInCalendarDays(startOfDay(to), startOfDay(from));
}

function isWithin(target: Date, start: Date, end: Date): boolean {
  const t = startOfDay(target).getTime();
  return t >= startOfDay(start).getTime() && t <= startOfDay(end).getTime();
}

// ---------------------------------------------------------------------------
// Main status calculation
// ---------------------------------------------------------------------------

/**
 * Calculate the full cycle status from period logs.
 * Returns undefined if no logs are provided.
 * Mirrors CycleCalculations.calculateCycleStatus(_:settings:referenceDate:) exactly.
 */
export function calculateCycleStatus(
  periodLogs: PeriodLog[],
  settings: CycleSettings,
  referenceDate: Date = new Date()
): CycleStatusResult | undefined {
  if (periodLogs.length === 0) return undefined;

  const ref = startOfDay(referenceDate);

  // Sort most-recent first (mirrors Swift: sorted { $0.startDate > $1.startDate })
  const sorted = [...periodLogs].sort(
    (a, b) => b.startDate.getTime() - a.startDate.getTime()
  );

  let cycleStartDate: Date | undefined;

  for (const period of sorted) {
    const pStart = startOfDay(period.startDate);
    const pEnd = period.endDate
      ? startOfDay(period.endDate)
      : addDays(pStart, settings.averagePeriodLengthDays - 1);

    if (isWithin(ref, pStart, pEnd)) {
      cycleStartDate = pStart;
      break;
    }
    const nextExpected = addDays(pStart, settings.averageCycleLengthDays);
    if (ref.getTime() > pEnd.getTime() && ref.getTime() < startOfDay(nextExpected).getTime()) {
      cycleStartDate = pStart;
      break;
    }
  }

  let cycleStart: Date;
  if (cycleStartDate !== undefined) {
    cycleStart = cycleStartDate;
  } else {
    // Fast-forward from most recent period by complete cycle lengths
    const most = sorted[0];
    let start = startOfDay(most.startDate);
    const daysSince = daysBetween(start, ref);
    const safeCycleLength = Math.max(1, settings.averageCycleLengthDays);
    const completed = Math.floor(daysSince / safeCycleLength);
    start = addDays(start, completed * safeCycleLength);
    cycleStart = start;
  }

  const cycleDay = daysBetween(cycleStart, ref) + 1;
  const boundaries = getPhaseBoundaries(settings);

  let currentPhase: CyclePhase = 'follicular';
  let phaseStartDay = 1;
  let phaseEndDay = settings.averageCycleLengthDays;

  for (const phase of PHASE_ORDER) {
    const b = boundaries[phase];
    if (cycleDay >= b.start && cycleDay <= b.end) {
      currentPhase = phase;
      phaseStartDay = b.start;
      phaseEndDay = b.end;
      break;
    }
  }

  let daysUntilNext: number;
  switch (currentPhase) {
    case 'menstrual':
      daysUntilNext = (boundaries.follicular.start) - cycleDay;
      break;
    case 'follicular':
      daysUntilNext = (boundaries.ovulation.start) - cycleDay;
      break;
    case 'ovulation':
      daysUntilNext = (boundaries.luteal.start) - cycleDay;
      break;
    case 'luteal':
      daysUntilNext = settings.averageCycleLengthDays - cycleDay + 1;
      break;
  }

  return {
    currentPhase,
    cycleDay,
    daysUntilNextPhase: Math.max(0, daysUntilNext),
    predictedNextPeriod: addDays(cycleStart, settings.averageCycleLengthDays),
    phaseStartDate: addDays(cycleStart, phaseStartDay - 1),
    phaseEndDate: addDays(cycleStart, phaseEndDay - 1),
  };
}

// ---------------------------------------------------------------------------
// Phase recommendation (informational — not used in adaptation pipeline)
// ---------------------------------------------------------------------------

export interface PhaseRecommendation {
  phase: CyclePhase;
  title: string;
  description: string;
  trainingFocus: string;
  intensityRecommendation: string;
  exercisesToEmphasize: string[];
  exercisesToAvoid: string[];
}

export function getPhaseRecommendation(phase: CyclePhase): PhaseRecommendation {
  switch (phase) {
    case 'menstrual':
      return {
        phase,
        title: 'Menstrual Phase',
        description: 'Your period phase. Energy may be lower — you might feel more fatigued.',
        trainingFocus: 'Recovery and light movement',
        intensityRecommendation: 'low',
        exercisesToEmphasize: ['yoga', 'walking', 'light stretching'],
        exercisesToAvoid: ['heavy compound lifts', 'max effort attempts'],
      };
    case 'follicular':
      return {
        phase,
        title: 'Follicular Phase',
        description: 'Energy and endurance begin to rise. Estrogen increases, supporting muscle growth.',
        trainingFocus: 'Building strength and endurance',
        intensityRecommendation: 'moderate',
        exercisesToEmphasize: ['compound movements', 'strength training', 'cardio'],
        exercisesToAvoid: [],
      };
    case 'ovulation':
      return {
        phase,
        title: 'Ovulation Phase',
        description: 'Peak estrogen and testosterone. Often the strongest phase for performance.',
        trainingFocus: 'High-intensity training and PR attempts',
        intensityRecommendation: 'peak',
        exercisesToEmphasize: ['max effort attempts', 'heavy compound lifts', 'power-focused workouts'],
        exercisesToAvoid: [],
      };
    case 'luteal':
      return {
        phase,
        title: 'Luteal Phase',
        description: 'Progesterone rises, which may affect recovery and energy. Focus on maintenance.',
        trainingFocus: 'Maintenance and technique refinement',
        intensityRecommendation: 'moderate',
        exercisesToEmphasize: ['technique work', 'volume training', 'recovery-focused sessions'],
        exercisesToAvoid: ['max effort attempts', 'extremely heavy loads'],
      };
  }
}
